reactioncommerce/reaction
# Shippo address validation

Resolves #3823 
Impact: **critical**  
Type: **bugfix**

## Issue
When a user saves an invalid address in the address book form while checking out, Shippo is unable to find any rates and ambiguously fails because they require the address to be valid to get rates. The error returned is "No shipping methods are configured".

## Solution
Aside from obviously making that error more accurate, Shippo also needs to have the address validated _before_ getting rates is attempted. They provide a separate API endpoint for this purpose. This PR adds server side methods that use this API to either confirm validity or return a better suggested address.
So basically Shippo works the same as Avalara.

## Breaking changes
None expected.

## Testing
1. Add a product to cart.
1. Go to the Checkout Screen.
1. Try entering a partially valid address. 
1. You should get a toast, regarding the error in the address.
1. Fix it, enter a almost complete address.
1. You should be taken to the address review screen. This screen has the same functionality as Avalara address review screen.
1. Try `Edit entered address`, and submit with same address. (Basically testing all the workflows work)
1. Try `Use selected address`, the address should be added and tax should be calculated.

Repeat the above strep with Shippo disabled and Avalara enabled(to check nothing is broken for Avalara).

## Help
```js
// simply provide the cartId and the method will grab the shipping address
Meteor.call("shippo/validateCartAddress", cartId, (err, result) => {
  if (!err) {
    console.log(result);
  }
});

// if valid
{    
  "object_state": "VALID",
  "object_purpose": "QUOTE",
  "object_source": "VALIDATOR",
  "object_created": "2018-03-28T13:21:19.090Z",
  "object_updated": "2018-03-28T13:21:19.370Z",
  "object_id": "3d9599e77b474de0b188e834b6e93209",
  "object_owner": "admin@reactioncommerce.com",
  "name": "Jeremy Shimko",
  "company": "Reaction Commerce",
  "street_no": "",
  "street1": "123 Main St",
  "street2": "",
  "street3": "",
  "city": "Burlington",
  "state": "CT",
  "zip": "06013-1513",
  "country": "US",
  "longitude": -72.93827,
  "latitude": 41.7792,
  "phone": "",
  "email": "me@example.com",
  "is_residential": true,
  "ip": "",
  "messages": [],
  "metadata": "",
  "test": true
}

// if invalid
{
  "object_state": "INVALID",
  "object_purpose": "QUOTE",
  "object_source": "VALIDATOR",
  "object_created": "2018-03-28T13:26:14.105Z",
  "object_updated": "2018-03-28T13:26:14.636Z",
  "object_id": "1ad6b9337a96493c8e39b6e9c9383512",
  "object_owner": "admin@reactioncommerce.com",
  "name": "Jeremy Shimko",
  "company": "Reaction Commerce",
  "street_no": "",
  "street1": "whatever st",
  "street2": "",
  "street3": "",
  "city": "STUFF",
  "state": "AK",
  "zip": "12345",
  "country": "US",
  "longitude": null,
  "latitude": null,
  "phone": "",
  "email": "me@example.com",
  "is_residential": null,
  "ip": "",
  "messages": [
    {
      "source": "Shippo Address Validator",
      "code": "Invalid City",
      "type": "address_error",
      "text": "The City in the address submitted is invalid. Remember, city names cannot begin with numbers."
    }
  ],
  "metadata": "",
  "test": true
}
```
